segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,510 @@
+        Map<String, TransactionData> txnData = txn.getData();
+        String key = metadata.getKey();
+
+        TransactionData data = TransactionData.builder().key(key).build();
+        TransactionData oldData = txnData.putIfAbsent(key, data);
+        if (null != oldData) {
+            data = oldData;
+        }
+
+        data.setValue(metadata);
+        data.setPinned(true);
+        data.setVersion(txn.getVersion());
+
+        pinnedKeys.add(metadata.getKey());
+    }
+
+    /**
+     * Creates a new metadata record.
+     *
+     * @param txn Transaction.
+     * @param metadata  metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void create(MetadataTransaction txn, StorageMetadata metadata) throws StorageMetadataException  {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != metadata);
+        Preconditions.checkArgument(null != metadata.getKey());
+        Map<String, TransactionData> txnData = txn.getData();
+        txnData.put(metadata.getKey(), TransactionData.builder()
+                .key(metadata.getKey())
+                .value(metadata)
+                .version(txn.getVersion())
+                .build() );
+    }
+
+
+    /**
+     * Deletes a metadata record given the key.
+     *
+     * @param txn Transaction.
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void delete(MetadataTransaction txn, String key) throws StorageMetadataException {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != key);
+        Map<String, TransactionData> txnData = txn.getData();
+
+        TransactionData data = TransactionData.builder().key(key).build();
+        TransactionData oldData = txnData.putIfAbsent(key, data);
+        if (null != oldData) {
+            data = oldData;
+        }
+        data.setValue(null);
+        data.setPersisted(false);
+        data.setVersion(txn.getVersion());
+    }
+
+    /**
+     * {@link AutoCloseable#close()} implementation.
+     */
+    @Override
+    public void close() throws Exception {
+        ArrayList<TransactionData> modifiedValues = new ArrayList<>();
+        bufferedTxnData.entrySet().stream().filter(entry -> !entry.getValue().isPersisted() && !entry.getValue().isPinned()).forEach(entry -> modifiedValues.add(entry.getValue()));
+        if (modifiedValues.size() > 0) {
+            writeAll(modifiedValues);
+        }
+    }
+
+    /**
+     * Marks the store as fenced.
+     */
+    public void markFenced() {
+        this.fenced.set(true);
+    }
+
+    /**
+     * Stores the transaction data.
+     */
+    @Builder(toBuilder = true)
+    @Data
+    public static class TransactionData {
+        /**
+         * Version
+         */
+        private long version;
```

Comment:
final everywhere

Reply:
fixed

Reply:
I don't think so.

Reply:
This field is changes during lifetime of this object.